### PR TITLE
Replace the Person class with Subject

### PR DESCRIPTION
**Pull request by hokwaichan**
# Ticket Link

[1654](https://uhawaii.atlassian.net/browse/GROUPINGS-1654)

# List of squashed commits

- Refactor Subject

# Test Checklist

- [x] Exhibits Clear Code Structure:
- [x] Project Unit Tests Passed:
- [x] Project Integration Tests Passed:
- [x] Executes Expected Functionality:
- [x] Tested For Incorrect/Unexpected Inputs:

### CODE REVIEW

**Comment by JorWo** [comment]
There are some Codacy issues to fix.

**Comment by JorWo** [comment]
Some merge conflicts to fix too, thanks.

File: src/main/java/edu/hawaii/its/api/access/UhCasAttributes.java [hunk 1]
```diff
@@ -52,36 +52,8 @@ public String getName() {
...
-
-                // Couldn't match up username with uid,
-                // so just return first value.
-                return values.get(0); // We are done.
-            }
-        }
-
-        return ""; // Didn't find anything.
-    }
+    public String getUid() { return getValue("uid"); }
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/access/UhCasAttributes.java:56]: Expand this line so that:
```
public String getUid() {
    return getValue("uid"); 
}
```
IntelliJ usually formats it like that but the actual code is actually expanded, thanks!

File: src/main/resources/application.properties [hunk 2]
```diff
@@ -65,11 +65,11 @@ groupings.api.failure=FAILURE
...
-groupings.api.person_attributes.username=uid
-groupings.api.person_attributes.uhuuid=uhUuid
-groupings.api.person_attributes.first_name=givenName
-groupings.api.person_attributes.last_name=sn
-groupings.api.person_attributes.composite_name=cn
+groupings.api.subject_attributes.uid=uid
+groupings.api.subject_attributes.uhuuid=uhUuid
+groupings.api.subject_attributes.first_name=givenName
+groupings.api.subject_attributes.last_name=sn
+groupings.api.subject_attributes.composite_name=cn
```
Comment (JorWo) [inline src/main/resources/application.properties:72]: I think these are not being used and can be removed. However, if you think there would be a good place to use these properties please use them.

File: src/main/resources/data.harness.properties [hunk 3]
```diff
@@ -12,36 +12,36 @@ find.groups.results.empty.description={"groupResults":[{"idIndex":"843904","type
...
 ws.subject.subject.uid.not.found={"resultCode":"SUBJECT_NOT_FOUND","success":"F","identifierLookup":"invalid-uid","id":null,"name":null,"sourceId":null,"attributeValues":null}
 ws.subject.subject.uhuuid.not.found={"resultCode":"SUBJECT_NOT_FOUND","success":"F","identifierLookup":null,"id":"11111111","name":null,"sourceId":null,"attributeValues":null}
-ws.subject.success.null.values={ "resultCode": null, "success": "T", "identifierLookup": "testiwta", "id": "99997010", "name": "Testf-iwt-a TestIAM-staff", "sourceId": "UH core LDAP", "attributeValues": [ "testiwta", "Testf-iwt-a TestIAM-staff", null, null, "", "eduPersonOrgDN=uhsystem,eduPersonAffiliation=staff", "staff", "uhsystem", "" ] }
+ws.subject.success.null.values={ "resultCode": null, "success": "T", "identifierLookup": "testiwta", "id": "99997010", "name": "Testf-iwt-a TestIAM-staff", "sourceId": "UH core LDAP", "attributeValues": [ "testiwta", "Testf-iwt-a TestIAM-staff", null, null, "", "eduSubjectOrgDN=uhsystem,eduSubjectAffiliation=staff", "staff", "uhsystem", "" ] }
```
Comment (JorWo) [inline src/main/resources/data.harness.properties:19]: These lines will have to stay as `eduPersonOrgDN` and `eduPersonAffiliation` because they come directly from Grouper and cannot be changed.

File: uh-groupings-api-overrides.skeleton.properties
Comment (JorWo) [inline uh-groupings-api-overrides.skeleton.properties:1]: Keep this file as username, new members would be used to UH username.

File: uh-groupings-api-spec-v2-2.yml
Comment (JorWo) [inline uh-groupings-api-spec-v2-2.yml:1]: You can undo the changes in this .yml file, it's very outdated a deserves an update.

File: src/test/resources/grouper.test.properties
Comment (JorWo) [inline src/test/resources/grouper.test.properties:1]: Please undo the changes in here.

File: src/test/resources/ootb.grouper.test.properties
Comment (JorWo) [inline src/test/resources/ootb.grouper.test.properties:1]: Please undo the changes in here. Like mentioned before, I will make a ticket to update property files for usernames and uh numbers to uid and uhUuid.

File: src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1.java [hunk 4]
```diff
@@ -416,7 +416,7 @@ public ResponseEntity<GroupingRemoveResults> removeExcludeMembers(
     }
 
     /**
-     * Get an owner's owned groupings by username or UH id number.
+     * Get an owner's owned groupings by uid or UH id number.
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1.java:419]: UH id number can be replaced with uhUuid here, thanks.

File: src/main/resources/grouper.client.base.properties [hunk 5]
```diff
@@ -525,7 +525,7 @@ esb.consumer.encryptionImplementation = edu.internet2.middleware.grouperClient.e
 ## general xmpp configuration
 grouperClient.xmpp.server.host = jabber.school.edu
 grouperClient.xmpp.server.port = 5222
-grouperClient.xmpp.user = username
+grouperClient.xmpp.user = uid
```
Comment (JorWo) [inline src/main/resources/grouper.client.base.properties:528]: Not sure if I commented this before, must've been on the UI PR, but revert this change back to username because this affects a different scope.

File: src/test/java/edu/hawaii/its/api/access/UhCasAttributesTest.java [hunk 6]
```diff
@@ -208,7 +205,7 @@ public void misc() {
         map.put("uhUuid", "666666");
         map.put("cn", "Frank");
         map.put("mail", "frank@example.com");
-        map.put("eduPersonAffiliation", "aff");
+        map.put("eduSubjectAffiliation", "aff");
```
Comment (JorWo) [inline src/test/java/edu/hawaii/its/api/access/UhCasAttributesTest.java:208]: The should still be eduPersonAffiliation, thanks.

**Comment by JorWo** [comment]
Also make sure to undo the changes in `application-integrationTest.properties` and `application-localTest.properties`.

File: src/test/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1Test.java [hunk 7]
```diff
@@ -136,25 +136,64 @@ private Grouping grouping() {
...
 
         Group basisGroup = new Group();
-        basisGroup.addMember(new Person("b0-name", "b0-uuid", "b0-username"));
-        basisGroup.addMember(new Person("b1-name", "b1-uuid", "b1-username"));
-        basisGroup.addMember(new Person("b2-name", "b2-uuid", "b2-username"));
-        grouping.setBasis(basisGroup);
+        Subject subjectBasis0 = new Subject();
+        subjectBasis0.setUid("b0-uid");
+        subjectBasis0.setName("b0-name");
+        subjectBasis0.setUhUuid("b0-uuid");
```
Comment (JorWo) [inline src/test/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1Test.java:142]: Maybe adding a constructor would be worth it here. Let me know if there's any difficulties with making the constructor. I know that the WsSubject class is hard to deal with.

File: src/test/java/edu/hawaii/its/api/groupings/GroupingMemberTest.java [hunk 8]
```diff
@@ -25,8 +25,8 @@
 @SpringBootTest(classes = { SpringBootWebApplication.class })
 public class GroupingMemberTest {
 
-    @Value("${groupings.api.test.uh-usernames}")
-    private List<String> TEST_USERNAMES;
+    @Value("${groupings.api.test.uh-uids}")
+    private List<String> TEST_UIDS;
```
Comment (JorWo) [inline src/test/java/edu/hawaii/its/api/groupings/GroupingMemberTest.java:29]: Sorry I am having you undo a lot these changes. I am trying to reduce the number of changes and split the remaining changes into another ticket. That way any possible merge conflicts are easier to deal with.

File: src/test/java/edu/hawaii/its/api/groupings/GroupingAddResultTest.java [hunk 9]
```diff
@@ -27,7 +27,7 @@
 public class GroupingAddResultTest {
 
     @Value("${groupings.api.test.uh-usernames}")
-    private List<String> TEST_USERNAMES;
+    private List<String> TEST_UIDS;
```
Comment (JorWo) [inline src/test/java/edu/hawaii/its/api/groupings/GroupingAddResultTest.java:30]: Final thing to change, please change TEST_UIDS back to TEST_USERNAMES for all test classes. This will be covered in another ticket, thanks!

**Comment by fduckart** [comment]
I don't think we can change the UhCasAttributes class from username to uid when the value it it's using is username. It's a little confusing at UH since a lot of times the username and the uid are the same, but Spring Security is using something it calls "username." They define the "username" as the value you enter in username field on the login page. One thing that might make this go away, though, is whether we even use the UhCasAttributes class? You could possibly just delete that whole class.

**Comment by hokwaichan** [comment]
I think we don't use the UhCasAttributes class for the API side, and we do have a ticket for deleting the UhCasAttributes class

**Review by JorWo — APPROVED** [review]
